node/network/protocol/src/peer_set.rs

```diff
@@ -181,6 +174,262 @@ impl<T> IndexMut<PeerSet> for PerPeerSet<T> {
+pub enum ValidationVersion {
+	/// The first version.
+	V1 = 1,
+}
+
+/// Supported collation protocol versions. Only versions defined here must be used in the codebase.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EnumIter)]
+pub enum CollationVersion {
+	/// The first version.
+	V1 = 1,
+}
+
+impl From<ValidationVersion> for ProtocolVersion {
+	fn from(version: ValidationVersion) -> ProtocolVersion {
+		version as ProtocolVersion
+	}
+}
+
+impl From<CollationVersion> for ProtocolVersion {
+	fn from(version: CollationVersion) -> ProtocolVersion {
+		version as ProtocolVersion
+	}
+}
+
+/// On the wire protocol name to [`PeerSet`] mapping.
+#[derive(Clone)]
+pub struct PeerSetProtocolNames {
+	protocols: HashMap<Cow<'static, str>, (PeerSet, ProtocolVersion)>,
+	names: HashMap<(PeerSet, ProtocolVersion), Cow<'static, str>>,
+}
+
+impl PeerSetProtocolNames {
+	/// Construct [`PeerSetProtocols`] using `genesis_hash` and `fork_id`.
+	pub fn new(genesis_hash: Hash, fork_id: Option<&str>) -> Self {
+		let mut protocols = HashMap::new();
+		let mut names = HashMap::new();
+		for protocol in PeerSet::iter() {
+			// All protocol versions from [`ValidationVersion`] and [`CollationVersion`] must be
+			// registered here. Currently it's only `MAIN_PROTOCOL_VERSION` = `V1` = 1.
+			// Register main protocol name.
+			let main_protocol_name =
+				Self::generate_name(&genesis_hash, fork_id, protocol, MAIN_PROTOCOL_VERSION);
+			names.insert((protocol, MAIN_PROTOCOL_VERSION), main_protocol_name.clone());
+			Self::insert_protocol_or_panic(
+				&mut protocols,
+				main_protocol_name,
+				protocol,
+				MAIN_PROTOCOL_VERSION,
+			);
+			// Register legacy protocol name.
+			Self::insert_protocol_or_panic(
+				&mut protocols,
+				Self::get_legacy_name(protocol),
+				protocol,
+				LEGACY_PROTOCOL_VERSION,
+			)
+		}
+		Self { protocols, names }
+	}
+
+	/// Helper function to make sure no protocols have the same name.
+	fn insert_protocol_or_panic(
+		protocols: &mut HashMap<Cow<'static, str>, (PeerSet, ProtocolVersion)>,
+		name: Cow<'static, str>,
+		protocol: PeerSet,
+		version: ProtocolVersion,
+	) {
+		match protocols.entry(name) {
+			Entry::Vacant(entry) => {
+				entry.insert((protocol, version));
+			},
+			Entry::Occupied(entry) => {
+				panic!(
+					"Protocol {:?} (version {}) has the same on-the-wire name as protocol {:?} (version {}): `{}`.",
+					protocol,
+					version,
+					entry.get().0,
+					entry.get().1,
+					entry.key(),
+				);
+			},
+		}
+	}
+
+	/// Lookup the protocol using its on the wire name.
+	pub fn try_get_protocol(&self, name: &Cow<'static, str>) -> Option<(PeerSet, ProtocolVersion)> {
+		self.protocols.get(name).map(ToOwned::to_owned)
+	}
+
+	/// Get the main protocol name. It's used by the networking for keeping track
+	/// of peersets and connections.
+	pub fn get_main_name(&self, protocol: PeerSet) -> Cow<'static, str> {
+		self.get_name(protocol, MAIN_PROTOCOL_VERSION)
+	}
```

Comment:
I would remove this. As there is no requirement that the validation & collation protocol have the same protocol version.

Reply:
"Main" protocol name (as opposed to negotiated protocol name) is needed, e.g., for peer sets bookkeeping, but I've specified validation & collation versions explicitly using new enums.